[RFR] Add API for Redfish chassis navigation

matejart opened the pull request:
We expand the Redfish API implementation to include retrieval and inspection of Chassis resources.

This PR introduces class `RedfishResource` that abstracts Redfish entities such as `Systems`, `Chassis` or `EventService`.

In the context of physical infrastructure, servers normally reside in one or more nested enclosures such as chassis or racks. This new entity serves as a reporesentation of these physical containers. Not to be confused by container virtualisation technology.

The Redfish API now supports retrieving Chassis resources, which are general chassis entities. It also provides helper methods for retrieving physical racks, which are a special type of Chassis. The Redfish racks currently support getting the name of the rack. The Redfish chassis support obtaining chassis name and description, status of the identifying LED (if available) and the number of servers installed within the chassis.

Related Bugzilla tickets:
  * https://bugzilla.redhat.com/show_bug.cgi?id=1571610
  * https://bugzilla.redhat.com/show_bug.cgi?id=1574809

matejart commented:
@miq-bot add_label Redfish

matejart commented:
/cc @miha-plesko @tadeboro

tadeboro commented on wrapanapi/entities/physical_container.py:
```diff
@@ -0,0 +1,30 @@
+from wrapanapi.entities.base import Entity
+
+
+class PhysicalContainer(Entity):
```
Maybe it would be nice to handle python 2/3 compatibility here since this is a new file? Something like this should work (if we import `six`, of course):

```suggestion
class PhysicalContainer(six.with_metaclass(ABCMeta, Entity)):
```

This way we can remove the python2-only `__metaclass__` assignment.

matejart replied:
👍

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -90,6 +63,58 @@ def uuid(self):
+    @property
+    def server_cores(self):
+        """Return the number of cores on this server."""
+        return sum([int(p.TotalCores) for p in self.raw.Processors.Members])
```
`[]` are not needed here and needlessly create new list. `sum` can work with iterators, so
```suggestion
        return sum(int(p.TotalCores) for p in self.raw.Processors.Members)
```
is enough.

matejart replied:
👍

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -178,7 +252,57 @@ def get_server(self, resource_id):
         """
         return RedfishServer(self, raw=self.api_client.find(resource_id))
 
+    def get_chassis(self, resource_id, required_types=None):
```
Maybe it would be a bit nicer for the method callers to define it as
```suggestion
    def get_chassis(self, resource_id, *required_types):
```

This will remove the need for adding `[]` around the last group of parameters.

matejart replied:
👍

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -103,7 +139,9 @@ class RedfishSystem(System):
     # statistics for the provider
     _stats_available = {
-        'num_server': lambda self: self.num_servers,
+        'num_server': lambda server: server.num_servers,
```
Nitpicking, but

```suggestion
        'num_server': lambda system: system.num_servers,
```

(same for the two lines below)

Because those functions are defined on RedfishSystem, not RedfishServer

matejart replied:
Well-spotted! 👍

mshriver replied:
This still doesn't look right to me - `RedfishChassis.num_servers` is where the method is defined, @miha-plesko referenced `System.num_servers`, but the diff still shows `server.num_servers` call?

miha-plesko replied:
Well actually the diff shown above is my "suggestion", a new feature of GitHub. You can try it in your comments with this button here:

![capture](https://user-images.githubusercontent.com/8102426/48420337-77560880-e75a-11e8-8103-23c0cb4e5c3e.PNG)

I see it just introduced confusion, sorry about that. As you say @mshriver the comment is still vaild - lambda argument should be renamed to 'server'.

mshriver replied:
Additionally, it looks like the two calls below to `num_chassis` and `num_racks` are incorrect as well?  Both of those methods are defined on the local RedfishSystem class, and should be called with `self.num_<thing>` yes?

matejart replied:
I really apologize for the confusion - this too was stuck in my un-pushed change. I belive that now it is as it should be - the `_stats_available` referring to `system`. 

As for the `rack` parameter in `_rack_*_available` and `chassis` parameter in `_chassis_*_available`, they get obtained through `get_rack` and `get_chassis`, respectively, so the properties are implemented in `RedfishChassis` (which represents both the Redfish racks and the Redfish chassis at the API end). The integration tests use the `RedfishSystem` just to get to racks, chassis, servers, etc., but the definition for the `num_<thing>`s is in their respective specialized classes.

matejart commented:
I addressed the comments and suggestions. I also expanded the functionality to now cover Redfish Chassis.
@miha-plesko can you please provide your review?

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -168,6 +225,82 @@ def server_inventory(self, physical_server, requested_items, **kwargs):
         return {item: self._server_inventory_available[item](redfish_server)
                 for item in requested_items}
 
+    def rack_stats(self, physical_rack, requested_stats, **kwargs):
```
Do we plan to use the `kwargs` or should we just omit it from the list?

matejart replied:
👍 removed.

mshriver replied:
`**kwargs` is still in the method definition, I'm starting to think that a commit from @matejart has not been pushed to this fork/PR?

matejart replied:
This is embarrassing. I in fact forgot to push these changes. The `**kwargs` are now gone.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -168,6 +225,82 @@ def server_inventory(self, physical_server, requested_items, **kwargs):
+        redfish_rack = self.get_rack(physical_rack.ems_ref)
+
+        return {stat: self._rack_stats_available[stat](redfish_rack)
+                for stat in requested_stats}
```
There's a bunch of boilerplate code in those 4 methods :) I'm thinking to refactor into something more like:

```python
def rack_stats(self, ...):
  return evaluate_lambdas(hash_with_lambdas, keys, lambda_args)

def rack_inventory(self, ...):
  return evaluate_lambdas(hash_with_lambdas, keys, lambda_args)

def chassis_stats(self, ...):
  return evaluate_lambdas(hash_with_lambdas, keys, lambda_args)

def chassis_inventory(self, ...):
  return evaluate_lambdas(hash_with_lambdas, keys, lambda_args)
```
(of course, you need to provide correct values for `hash_with_lambdas`, `keys` and `lambda_args`)

Because we're doing the same thing all the time, evaluating correct lambdas... Maybe we could put `evaluate_lambdas` in a Mixin even to make it reusable.

matejart replied:
Thank you for the suggestion. I created a [ticket](https://github.com/xlab-si/wrapanapi/issues/1) to address this later.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -178,7 +311,57 @@ def get_server(self, resource_id):
+           InvalidValueException if the resource_id represents a Chassis that is
+              not of any of the required types
+        """
+        chassis = RedfishChassis(self, raw=self.api_client.find(resource_id))
```
Out of curiosity: how does `self.api_client` know that we're after chassis and not some other entity? Is entity type part of ID itself?

Also: what happens when chasis by this ID is not found? Will api client raise exception or will we end up with malfunctioning `RedfishChassis` instance? I'd prefer if we return `None` rather than malfunctioning instance.

matejart replied:
We are relying on the caller of the method to know what it is using the instance for.
If the id points to no resource, `self.api_client.find` will throw an exception.

mshriver replied:
@matejart what type of exception is raised? I'm thinking it should be caught and `wrapanapi.exceptions.ItemNotFound` raised instead, or a new specific exception type added for `RedfishItemNotFound`.  That way caller has a consistent expectation for what the exception will be if their 'Item' resource_id isn't found, and it doesn't appear to be an exception in the core API.

matejart commented on wrapanapi/systems/redfish.py:
```diff
@@ -31,7 +28,53 @@ def __init__(self, system, raw=None, **kwargs):
+        These attributes identify the instance without needing to query the API
+        for updated data.
+        """
+        return {'ems_ref': self.name}
```
The problem is that Redfish doesn't have an uuid.

miha-plesko replied:
@matejart but I can see the `def uuid` property implemented in L66 ;)

mshriver replied:
This has been marked as resolved but @miha-plesko and @matejart you don't seem to be on the same page here. There is a UUID property defined, using the raw `Id` property. Why not use it here? If there is a valid reason, please include in-line in a comment why name is used.

matejart replied:
I am a bit unsure about what is the recommended practice in this context. For Redfish resource, the property `name` returns a unique idenfier of the instance. Considering this, would it make sense to return simply `{'name': self.name}`?
So unless `def uuid` is required, I would remove it because it is confusing and because a Redfish resource instance does not (necessarily) have a UUID.

But if I have to implement `def uuid`, then I will change as suggested to `{'uuid': self.uuid}` or `{'ems_ref': self.uuid}`.

matejart commented on wrapanapi/systems/redfish.py:
```diff
@@ -31,7 +28,53 @@ def __init__(self, system, raw=None, **kwargs):
+        Returns:
+            New value of self.raw
+        """
+        self.raw._cache = {}
```
No, the next request will populate everything automatically.

mshriver commented on wrapanapi/entities/physical_container.py:
```diff
@@ -0,0 +1,28 @@
+from wrapanapi.entities.base import Entity
+
+
+class PhysicalContainer(six.with_metaclass(ABCMeta, Entity)):
```
Would be nice to backport this to all the other entity derived classes for py2+py3 compatibility.

matejart replied:
👍 
I am only not sure how to test the changes.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -83,11 +56,74 @@ def refresh(self):
+    def num_servers(self):
+        """Retrieve the number of physical servers within this chassis."""
+        if "ComputerSystems" in self.raw.Links:
+            return len(self.raw.Links.ComputerSystems)
```
Considering using getattr to flatten this a bit:

```
return len(getattr(self.raw.Links, 'ComputerSystems', []))
```

matejart replied:
👍 
I did the same in `def led_state`.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -178,7 +311,57 @@ def get_server(self, resource_id):
+        Args:
+          resource_id: the Redfish @odata.id of the resource representing the
+             chassis to be retrieved
+          required_types: optional list of ChassisType values that the retrieved
```
Its not clear to me here if these are just strings, or class constants?

matejart replied:
They are strings. I changed the description to hopefully make it clearer.

matejart commented:
Thank you @mshriver for great comments. I have now addressed them.
